diff --git a/patch_cirrus.c b/patch_cirrus.c
index 4d60d66..f5af6fe 100644
--- a/patch_cirrus.c
+++ b/patch_cirrus.c
@@ -148,11 +148,11 @@ struct cs_spec {
 	int capture_init;
 
 	// new item to limit times we redo unmute/play
-	struct timespec last_play_time;
+	struct timespec64 last_play_time;
 	// record the first play time - we have a problem there
 	// some initial plays that I dont understand - so skip any setup
 	// till sometime after the first play
-	struct timespec first_play_time;
+	struct timespec64 first_play_time;
 
 };
 
@@ -1825,7 +1825,7 @@ void cs_8409_jack_unsol_event(struct hda_codec *codec, unsigned int res)
 
         mycodec_info(codec, "cs_8409_jack_unsol_event UNSOL 0x%08x tag 0x%02x\n",res,tag);
 
-        event = snd_hda_jack_tbl_get_from_tag(codec, tag);
+        event = snd_hda_jack_tbl_get_from_tag(codec, tag, 0);
         if (!event)
                 return;
         event->jack_dirty = 1;
@@ -2600,9 +2600,9 @@ static int patch_cs8409(struct hda_codec *codec)
        spec->capture_init = 0;
 
        // init the last play time
-       getnstimeofday(&(spec->last_play_time));
+       ktime_get_real_ts64(&(spec->last_play_time));
 
-       getnstimeofday(&(spec->first_play_time));
+       ktime_get_real_ts64(&(spec->first_play_time));
 
        myprintk("snd_hda_intel: Post Patching for CS8409\n");
        //mycodec_info(codec, "Post Patching for CS8409\n");
diff --git a/patch_cirrus_new84.h b/patch_cirrus_new84.h
index 910ce97..335b20a 100644
--- a/patch_cirrus_new84.h
+++ b/patch_cirrus_new84.h
@@ -1300,8 +1300,8 @@ static void cs_8409_pcm_playback_pre_prepare_hook(struct hda_pcm_stream *hinfo,
 	struct cs_spec *spec = codec->spec;
 
 	if (action == HDA_GEN_PCM_ACT_PREPARE) {
-		struct timespec curtim;
-		getnstimeofday(&curtim);
+		struct timespec64 curtim;
+		ktime_get_real_ts64(&curtim);
 		myprintk("snd_hda_intel: command cs_8409_pcm_playback_pre_prepare_hook HOOK PREPARE init %d last %ld cur %ld",spec->play_init,spec->last_play_time.tv_sec,curtim.tv_sec);
 		if (1) {
 			struct hda_cvt_setup *p = NULL;
@@ -1416,8 +1416,8 @@ static void cs_8409_playback_pcm_hook(struct hda_pcm_stream *hinfo, struct hda_c
 		myprintk("snd_hda_intel: command cs_8409_playback_pcm_hook open end");
 	} else if (action == HDA_GEN_PCM_ACT_PREPARE) {
 		// so this comes AFTER the stream format, frequency setup verbs are sent for the pcm stream
-		struct timespec curtim;
-		getnstimeofday(&curtim);
+		struct timespec64 curtim;
+		ktime_get_real_ts64(&curtim);
 		myprintk("snd_hda_intel: command cs_8409_playback_pcm_hook HOOK PREPARE init %d last %ld cur %ld",spec->play_init,spec->last_play_time.tv_sec,curtim.tv_sec);
 		//int power_chk = 0;
 		//power_chk = snd_hda_codec_read(codec, codec->core.afg, 0, AC_VERB_GET_POWER_STATE, 0);
